src/components/oauth/OAuthCodeHandler.js

```diff
@@ -0,0 +1,100 @@
+/**
+ * @author sarahr
+ */
+
+import React, { useCallback, useEffect, useState } from "react";
+import { useRouter } from "next/router";
+import { useQuery } from "react-query";
+import { build } from "@cyverse-de/ui-lib";
+
+import {
+    APP_CATEGORIES_QUERY_KEY,
+    getPrivateCategories,
+} from "serviceFacades/apps";
+import {
+    OAUTH_CALLBACK_QUERY_KEY,
+    doOAuthCallback,
+} from "serviceFacades/oauth";
+
+import OAuthErrorHandler from "./OAuthErrorHandler";
+import OAuthLinkingMessage from "./OAuthLinkingMessage";
+import ids from "./ids";
+import { getListingPath } from "components/apps/utils";
+import systemId from "components/models/systemId";
+import { useUserProfile } from "contexts/userProfile";
+
+const ERR_SERVICE = "general_service_error";
+
+function OAuthCodeHandler(props) {
+    const { baseId, apiName, code, stateId } = props;
+    const handlerId = build(baseId, ids.OAUTH_CODE_HANDLER);
+    const router = useRouter();
+    const [userProfile] = useUserProfile();
+    const [categoryQueryKey, setCategoryQueryKey] = useState([
+        APP_CATEGORIES_QUERY_KEY,
+        userProfile?.id,
+    ]);
+    const [listingUrl, setListingUrl] = useState(
+        getListingPath(null, null, null, null, null, null, false)
+    );
+
+    const determineListingUrl = useCallback(
+        (data) => {
+            const hpcCategory = data.categories.find(
+                (cat) => cat.system_id === systemId.agave
+            );
+            setListingUrl(
+                getListingPath(
+                    null,
+                    null,
+                    null,
+                    null,
+                    null,
+                    JSON.stringify(hpcCategory),
+                    false
+                )
+            );
+        },
+        [setListingUrl]
+    );
+
+    const { isFetching: isFetchingCategories } = useQuery({
+        queryKey: categoryQueryKey,
+        queryFn: getPrivateCategories,
+        config: {
+            onSuccess: determineListingUrl,
+            staleTime: Infinity,
+            cacheTime: Infinity,
+        },
+    });
+
+    useEffect(() => {
+        setCategoryQueryKey([APP_CATEGORIES_QUERY_KEY, userProfile?.id]);
+    }, [userProfile]);
```

Comment:
As an alternative to using state to manage the app category query key, the query's `config` can use the `enabled` flag based on the presence of the user's profile ID, and this query key array can be used directly in the `queryKey`.

Here's an example where the `enabled` flag is based on having a user ID (though it's not used in the query key here):
https://github.com/cyverse-de/sonora/blob/3d8cb7640ef34c6192d8559f8d33f7fbbc48931f/src/pages/apps/%5BsystemId%5D/%5BappId%5D/edit.js#L34-L44